### DISCUSSION TITLE
Cycle through three different actions with consecutive keypresses

### DISCUSSION THREAD

**Discussion by fooness** [Hammerspoon/hammerspoon#3507] [Q&A] [answered]
Hello everyone.

I would like to improve my window management keymaps to not only move/resize windows to, for example, the left *half* of the screen, but also the left one_third ord two_thirds. There’s probably some rather easy and very elegant solution to this, but I did not find anything helpful during my research.

Here’s my prototypical code, which … is definitely *not* very elegant.

Also I would like to implement the same behavior for other keymaps, so the `local i`, if necessary at all, should be local _inside_ the function, not outside as it is now.

```
local i

-- move / resize window left
hs.hotkey.bind({ "cmd", "ctrl" }, "h", function()
  if i == 0 then
    hs.window.focusedWindow():moveToUnit({ 0, 0, 0.5, 1 }, 0)
    i = i + 1
  elseif i == 1 then
    hs.window.focusedWindow():moveToUnit({ 0, 0, 0.33, 1 }, 0)
    i = i + 1
  elseif i == 2 then
    hs.window.focusedWindow():moveToUnit({ 0, 0, 0.67, 1 }, 0)
    i = i + 1
  else
    i = 0
  end
end)
```

Maybe someone could help out, please?

PS: I *could* use https://github.com/peterklijn/hammerspoon-shiftit for my purposes, but from a learning-perspective I’d still be interested if there’s an *easy* and elegant solution besides something rather complex as the following examples:

- https://stackoverflow.com/a/69194126
- https://stackoverflow.com/a/44350843

Or maybe, well, there *is* not easier solution to this problem …

**Comment by megalithic** [comment]
@fooness perhaps ask over on Discord too: https://discord.gg/3KxPBF9v

**Comment by muescha** [comment, marked as answer]
you can do this with a helper:

```lua
function hotkeybindcycle(modifier, key, message, callbackMap)
    local currentIndex = 1
    local maxIndex = #callbackMap
    local hotkeyHandler
    hotkeyHandler = hs.hotkey.bind(modifier, key, message, function()
        callbackMap[currentIndex]()
        currentIndex = (currentIndex % maxIndex) + 1
    end)
end
```
source: https://github.com/muescha/dot_hammerspoon/blob/master/Helpers/HotkeyBindCycle.lua


```lua
require("Helpers.HotkeyBindCycle")

fileInfo()

hotkeybindcycle(hyper, "v", keyInfo("Cycle abc"), {
    function() print("a") end,
    function() print("b") end,
    function() print("c") end,
})

hotkeybindcycle(hyper, "b", keyInfo("Cycle defg"), {
    function() print("d") end,
    function() print("e") end,
    function() print("f") end,
    function() print("g") end,
})
```
source: https://github.com/muescha/dot_hammerspoon/blob/master/Functions/HotkeybindCycleTest.lua


you need to adjust the `require` to your file structure - `fileInfo()` and `keyInfo()` you can remove and add your `message`,  `local hotkeyHandler` is on an extra line so you can access it and add `hotkeyHandler:disable()` and `hotkeyHandler:disable()` if there is need.

**Comment by muescha** [comment]
if you like to have an variable for all keys you can do a table with the key `{modifier,key}` or your own textkey/objectkey (and yes you can use `{}` objects as a key for a table) :

```lua
cycleIndex = {}

modifier = { "cmd", "ctrl" }
key = "h"

cycleIndex[{modifier,key}] = 1 -- init index 
hs.hotkey.bind(modifier,key, function() 
	var currentIndex = cycleIndex[{modifier,key}]  -- get the current index  
	if currentIndex == 1 then
    -- do something
	elseif currentIndex == 2 then
	-- do something
    elseif currentIndex == 3 then
	-- do something
	end
	cycleIndex[{modifier,key}] = (cycleIndex[{modifier,key}] % maxIndex) + 1
end)
```
